mobile/lib/shared/services/hash.service.dart

```diff
@@ -0,0 +1,183 @@
+        : _db.iOSDeviceAssets.getById(a.localId!));
+    if (deviceAsset != null) return deviceAsset.hash;
+    final file = await a.local!.originFile;
+    if (file == null) {
+      _log.warning("Failed to get file for asset ${a.id}, skipping");
+      return null;
+    }
+    final hash =
+        await (Platform.isAndroid ? _hashFile(file) : _hashAssetCrypto(file));
+    if (hash == null) {
+      _log.warning("Failed to hash file ${file.path} for asset ${a.id}");
+      return null;
+    }
+    if (Platform.isAndroid) {
+      final da = AndroidDeviceAsset(id: a.localId!.toInt(), hash: hash);
+      await _db.writeTxn(() => _db.androidDeviceAssets.put(da));
+    } else {
+      final da = IOSDeviceAsset(id: a.localId!, hash: hash);
+      await _db.writeTxn(() => _db.iOSDeviceAssets.put(da));
+    }
+    return hash;
+  }
+
+  Future<List<Asset>> hashAssets(List<AssetEntity> assets) async {
+    final Stopwatch totalTime = Stopwatch()..start();
+    const int batchFileCount = 128;
+    const int batchDataSize = 1024 * 1024 * 1024; // 1GB
+    int totalBytes = 0;
+    int totalFiles = 0;
+    final List<DeviceAsset?> hashes;
+    final Stopwatch sw = Stopwatch()..start();
+    final List<DeviceAsset> toAdd = [];
+    final ids = assets
+        .map(Platform.isAndroid ? (a) => a.id.toInt() : (a) => a.id)
+        .toList();
+    hashes = await (Platform.isAndroid
+        ? _db.androidDeviceAssets.getAll(ids.cast())
+        : _db.iOSDeviceAssets.getAllById(ids.cast()));
+    debugPrint("Checking hash DB took ${sw.elapsedMilliseconds}ms");
+    sw.reset();
+    int bytes = 0;
+    final List<String> toHash = [];
+    for (int i = 0; i < assets.length; i++) {
+      if (hashes[i] != null) continue;
+      final file = await assets[i].originFile;
+      if (file == null) {
+        _log.warning("Failed to get file for asset ${assets[i].id}, skipping");
+        continue;
+      }
+      bytes += file.lengthSync();
+      toHash.add(file.path);
+      final da = Platform.isAndroid
+          ? AndroidDeviceAsset(id: ids[i] as int, hash: const [])
+          : IOSDeviceAsset(id: ids[i] as String, hash: const []);
+      toAdd.add(da);
+      totalFiles++;
+      hashes[i] = da;
+      if (toHash.length == batchFileCount ||
+          bytes >= batchDataSize ||
+          (i + 1 == assets.length && toHash.isNotEmpty)) {
+        final h = await _hashFiles(toHash);
+        bool anyNull = false;
+        for (int j = 0; j < h.length; j++) {
+          if (h[j] != null) {
+            toAdd[j].hash = h[j]!;
+          } else {
+            _log.warning("Failed to hash file ${toHash[j]}, skipping");
+            anyNull = true;
+          }
+        }
+        debugPrint(
+          "Hashing ${toAdd.length} files / $bytes b took ${sw.elapsedMilliseconds}ms: ${(bytes / (1024 * 1024.0)) / sw.elapsedMilliseconds * 1000} mb/s",
+        );
+        final validHashes =
+            anyNull ? toAdd.where((e) => e.hash.isNotEmpty).toList() : toAdd;
```

Comment:
What is the purpose in tracking `anyNull`. It seems like a performance optimization to avoid filtering the list here. The code would probably be clearer if you deleted lines 102, 108, and updated this to just always filter out empty values.